src/main/java/org/embulk/input/marketo/delegate/ProgramInputPlugin.java

```diff
@@ -0,0 +1,224 @@
+                        throw new ConfigException(String.format("`latest_updated_at` (%s) cannot precede the current date (%s) when incremental import",
+                                        latest.toString(DATE_FORMATER),
+                                        (DateTime.now().toString(DATE_FORMATER))));
+                    }
+                }
+                else {
+                    // Update the latestUpdatedAt for the config
+                    DateTime latest = earliest.plus(task.getReportDuration().get());
+                    // Only import until now
+                    if (latest.isAfter(DateTime.now())) {
+                        latest = DateTime.now();
+                    }
+                    task.setLatestUpdatedAt(Optional.of(latest.toDate()));
+                }
+            }
+
+            DateTime latest = new DateTime(task.getLatestUpdatedAt().get());
+
+            if (earliest.isAfter(DateTime.now())) {
+                throw new ConfigException(String.format("`earliest_updated_at` (%s) cannot precede the current date (%s)",
+                                earliest.toString(DATE_FORMATER),
+                                (DateTime.now().toString(DATE_FORMATER))));
+            }
+
+            if (earliest.isAfter(latest)) {
+                throw new ConfigException(String.format("Invalid date range. `earliest_updated_at` (%s) cannot precede the `latest_updated_at` (%s).",
+                                earliest.toString(DATE_FORMATER),
+                                latest.toString(DATE_FORMATER)));
+            }
+            // if filter type is selected, filter value must be presented
+            if (task.getFilterType().isPresent() && (!task.getFilterValues().isPresent() || task.getFilterValues().get().isEmpty())) {
+                throw new ConfigException("filter_value is required when selected filter_type");
+            }
+        }
+    }
+
+    @Override
+    protected Iterator<ServiceRecord> getServiceRecords(MarketoService marketoService, PluginTask task)
+    {
+        Iterable<ObjectNode> nodes = null;
+        if (!task.getQueryBy().isPresent()) {
+            nodes = marketoService.getPrograms();
+        }
+        else if (task.getQueryBy().isPresent() && task.getQueryBy().get() == QueryBy.TAG_TYPE) {
+            nodes = marketoService.getProgramsByTag(task.getTagType().get(), task.getTagVallue().get());
+        }
+        else if (task.getQueryBy().isPresent() && task.getQueryBy().get() == QueryBy.DATE_RANGE) {
+            nodes = marketoService.getProgramsByDateRange(task.getEarliestUpdatedAt().get(),
+                            task.getLatestUpdatedAt().get(),
+                            task.getFilterType().orNull(),
+                            task.getFilterValues().orNull());
+        }
+
+        return FluentIterable.from(nodes).transform(MarketoUtils.TRANSFORM_OBJECT_TO_JACKSON_SERVICE_RECORD_FUNCTION).iterator();
```

Comment:
Probably because of validation in `validateInputTask` but there is still room for nodes to be null which means 3 above conditions do not match and `nodes` is `null`, we should update to meaningful default value which will not bring any harmful if no condition matches.